src/code/audio/audio_effects.c

```diff
@@ -1,21 +1,340 @@
+    f32 channelVolume;
+    f32 chanFreqScale;
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80196BC8.s")
+    if (channel->changes.s.volume || recalculateVolume) {
+        channelVolume = channel->volume * channel->volumeScale * channel->seqPlayer->appliedFadeVolume;
+        if (channel->seqPlayer->muted && (channel->muteBehavior & 0x20)) {
+            channelVolume = channel->seqPlayer->muteVolumeScale * channelVolume;
+        }
+        channel->appliedVolume = channelVolume * channelVolume;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80196D20.s")
+    if (channel->changes.s.pan) {
+        channel->pan = channel->newPan * channel->panChannelWeight;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80196D7C.s")
+    chanFreqScale = channel->freqScale;
+    if (applyBend) {
+        chanFreqScale *= channel->seqPlayer->bend;
+        channel->changes.s.freqScale = true;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80196DB4.s")
+    for (i = 0; i < ARRAY_COUNT(channel->layers); i++) {
+        SequenceLayer* layer = channel->layers[i];
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80196FEC.s")
+        if (layer != NULL && layer->enabled && layer->note != NULL) {
+            if (layer->notePropertiesNeedInit) {
+                layer->noteFreqScale = layer->freqScale * chanFreqScale;
+                layer->noteVelocity = layer->velocitySquare2 * channel->appliedVolume;
+                layer->notePan = (channel->pan + layer->pan * (0x80 - channel->panChannelWeight)) >> 7;
+                layer->notePropertiesNeedInit = false;
+            } else {
+                if (channel->changes.s.freqScale) {
+                    layer->noteFreqScale = layer->freqScale * chanFreqScale;
+                }
+                if (channel->changes.s.volume || recalculateVolume) {
+                    layer->noteVelocity = layer->velocitySquare2 * channel->appliedVolume;
+                }
+                if (channel->changes.s.pan) {
+                    layer->notePan = (channel->pan + layer->pan * (0x80 - channel->panChannelWeight)) >> 7;
+                }
+            }
+        }
+    }
+    channel->changes.asByte = 0;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80197048.s")
+void AudioEffects_SequencePlayerProcessSound(SequencePlayer* seqPlayer) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80197138.s")
+    if ((seqPlayer->fadeTimer != 0) && (seqPlayer->skipTicks == 0)) {
+        seqPlayer->fadeVolume += seqPlayer->fadeVelocity;
+        seqPlayer->recalculateVolume = true;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_effects/func_80197164.s")
+        if (seqPlayer->fadeVolume > 1.0f) {
+            seqPlayer->fadeVolume = 1.0f;
+        }
+        if (seqPlayer->fadeVolume < 0) {
+            seqPlayer->fadeVolume = 0;
+        }
```

Comment:
I'm assuming a full `CLAMP` doesn't work, but maybe a `CLAMP_MAX` followed by a `CLAMP_MIN`?

Reply:
I gave all combinations of `CLAMP`, `CLAMP_MAX`, and `CLAMP_MIN` a shot but nothing worked. That's usually the case when it comes to floats, as floats tend to not be able to optimize out the extra `seqPlayer->fadeVolume = seqPlayer->fadeVolume` that comes with any form of clamps.

I did change `0` to `0.0f` since it is a float though